### PR TITLE
ENG-12281 The leader assignment improvement upon nodes down

### PR DESCRIPTION
**Pull request by zeemanhe**
improve the new leader assignment upon nodes down. A new host for partition leader is chosen based on the partition leader distributions among the nodes.

### CODE REVIEW

**Comment by zeemanhe** [comment]
Test case https://github.com/VoltDB/pro/pull/1887

File: src/frontend/org/voltdb/iv2/LeaderAppointer.java [hunk 1]
```diff
@@ -251,10 +258,30 @@ public void run(ImmutableMap<Integer, Long> cache) {
                     VoltDB.crashLocalVoltDB("Failed to get partition count", true, e);
                 }
             }
+            Set<Integer> liveHosts = Sets.newHashSet();
+            if (m_messenger != null){
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/iv2/LeaderAppointer.java]: Is it possible for `m_messenger ` to be NULL at any time?

File: src/frontend/org/voltdb/iv2/LeaderAppointer.java [hunk 2]
```diff
@@ -686,4 +734,59 @@ public void run() {
...
+
+        ImmutableSortedMap<Integer, Host> masterCountByHost = m_hostInfoMap.get();
+        Set<Integer> liveHosts = Sets.newHashSet();
+        if (m_messenger != null){
+            liveHosts = m_messenger.getLiveHostIds();
+        }
+        //check newly joined hosts.If a newly joined host has the replica,
+        //put the master on it since a newly joined host does not have masters on it
+        for (Integer hostId : liveHosts) {
+            if (candidateHost.contains(hostId) && !masterCountByHost.containsKey(hostId)) {
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/iv2/LeaderAppointer.java]: Is it ever possible for `candidateHost ` to be on a dead host?

File: src/frontend/org/voltdb/iv2/LeaderAppointer.java [hunk 3]
```diff
@@ -686,4 +734,59 @@ public void run() {
...
+            candidateHost.add(CoreUtils.getHostIdFromHSId(site));
+        }
+
+        ImmutableSortedMap<Integer, Host> masterCountByHost = m_hostInfoMap.get();
+        Set<Integer> liveHosts = Sets.newHashSet();
+        if (m_messenger != null){
+            liveHosts = m_messenger.getLiveHostIds();
+        }
+        //check newly joined hosts.If a newly joined host has the replica,
+        //put the master on it since a newly joined host does not have masters on it
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/iv2/LeaderAppointer.java]: Is it true if in future our automatic leader rebalance is introduced?

File: src/frontend/org/voltdb/iv2/LeaderAppointer.java [hunk 4]
```diff
@@ -686,4 +734,59 @@ public void run() {
...
+        for (Host host : hosts) {
+            if (candidateHost.contains(host.id)) {
+                host.increasePartitionLeader();
+                if (tmLog.isDebugEnabled()) {
+                    tmLog.debug("[findHostForPartitionLeader]: found host " + host.id + " for partition " + partitionId);
+                }
+                return host.id;
+            }
+        }
+        return newMasterHost;
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/iv2/LeaderAppointer.java]: Declared variable `newMasterHost ` is not used until the last return. Please return "-1" directly with an ENUM to make it understood easier.

File: src/frontend/org/voltdb/iv2/LeaderAppointer.java [hunk 5]
```diff
@@ -686,4 +734,59 @@ public void run() {
...
+            builder.append("\nHost Info:" + hosts);
+            tmLog.debug(builder.toString());
+        }
+        for (Host host : hosts) {
+            if (candidateHost.contains(host.id)) {
+                host.increasePartitionLeader();
+                if (tmLog.isDebugEnabled()) {
+                    tmLog.debug("[findHostForPartitionLeader]: found host " + host.id + " for partition " + partitionId);
+                }
+                return host.id;
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/iv2/LeaderAppointer.java]: Will the new leader update the `m_hostInfoMap` once it has been assigned? If not, how could the second partition assignment get the updated view of current leaders?

File: src/frontend/org/voltdb/iv2/LeaderAppointer.java [hunk 6]
```diff
@@ -237,11 +243,12 @@ public void run(List<String> children)
         @Override
         public void run(ImmutableMap<Integer, Long> cache) {
             Set<Long> currentLeaders = new HashSet<Long>(cache.values());
-            tmLog.debug("Updated leaders: " + currentLeaders);
             if (m_state.get() == AppointerState.CLUSTER_START) {
                 try {
                     if (currentLeaders.size() == getInitialPartitionCount()) {
-                        tmLog.debug("Leader appointment complete, promoting MPI and unblocking.");
+                        if (tmLog.isDebugEnabled()) {
```
Comment (zeemanhe) [inline src/frontend/org/voltdb/iv2/LeaderAppointer.java]: will remove the check

File: src/frontend/org/voltdb/iv2/LeaderAppointer.java (hunk 1 above)
Comment (zeemanhe) [inline src/frontend/org/voltdb/iv2/LeaderAppointer.java]: will remove the check

File: src/frontend/org/voltdb/iv2/LeaderAppointer.java [hunk 7]
```diff
@@ -251,10 +258,30 @@ public void run(ImmutableMap<Integer, Long> cache) {
...
+                    Host host = hostLeaderMap.get(hostId);
+                    if (host == null) {
+                        host = new Host(hostId);
+                        hostLeaderMap.put(hostId, host);
+                    }
+                    host.increasePartitionLeader();
+                }
+            }
+            m_hostInfoMap.set(ImmutableSortedMap.copyOf(hostLeaderMap));
+            if (tmLog.isDebugEnabled()) {
```
Comment (zeemanhe) [inline src/frontend/org/voltdb/iv2/LeaderAppointer.java]: The string construction has perf overhead

File: src/frontend/org/voltdb/iv2/LeaderAppointer.java (hunk 3 above)
Comment (zeemanhe) [inline src/frontend/org/voltdb/iv2/LeaderAppointer.java]: yeah.

File: src/frontend/org/voltdb/iv2/LeaderAppointer.java (hunk 2 above)
Comment (zeemanhe) [inline src/frontend/org/voltdb/iv2/LeaderAppointer.java]: candidateHost, from Zk callback, should not contain dead hosts.

**Comment by zeemanhe** [comment]
The leader determination algo has been improved.  New leaders are pre-calculated before partition callbacks are triggered so that partition callbacks share the same topology view.

File: src/frontend/org/voltdb/AbstractTopology.java [hunk 8]
```diff
@@ -1532,4 +1533,49 @@ public static AbstractTopology mutateRecoverTopology(AbstractTopology topology,
...
         return convertMutablesToTopology(topology.version, mutableHostMap, mutablePartitionMap);
     }
+
+    /**
+     * build AbstractTopology from @Statistics TOPO
+     * @param vt from @Statistics TOPO
+     * @param kfactor kfactor
+     * @return  AbstractTopology
+     */
+    public static AbstractTopology topologyFromStatistics(VoltTable vt, int kfactor) {
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/AbstractTopology.java]: Not used.

**Pull request merged**